src/channel.cr

```diff
@@ -102,82 +114,126 @@ class Channel(T)
     Closed
   end
 
-  private record Sender(T), fiber : Fiber, value : T, state_ptr : DeliveryState*, select_context : SelectContext(Nil)?
-  private record Receiver(T), fiber : Fiber, value_ptr : T*, state_ptr : DeliveryState*, select_context : SelectContext(T)?
+  # :nodoc:
+  struct Sender(T)
+    include Crystal::PointerLinkedList::Node
+    property fiber : Fiber
+    property data : T
+    property state : DeliveryState
+    property select_context : SelectContext(Nil)?
+
+    def initialize
+      @fiber = uninitialized Fiber
+      @data = uninitialized T
+      @state = DeliveryState::None
+    end
+  end
+
+  # :nodoc:
+  struct Receiver(T)
+    include Crystal::PointerLinkedList::Node
+    property fiber : Fiber
+    property data : T
+    property state : DeliveryState
+    property select_context : SelectContext(T)?
+
+    def initialize
+      @fiber = uninitialized Fiber
+      @data = uninitialized T
+      @state = DeliveryState::None
+    end
+  end
 
   def initialize(@capacity = 0)
     @closed = false
-    @senders = Deque(Sender(T)).new
-    @receivers = Deque(Receiver(T)).new
+
+    @senders = Crystal::PointerLinkedList(Sender(T)).new
+    @receivers = Crystal::PointerLinkedList(Receiver(T)).new
+
     if capacity > 0
       @queue = Deque(T).new(capacity)
     end
   end
 
-  def close
+  def close : Nil
+    sender_list = Crystal::PointerLinkedList(Sender(T)).new
+    receiver_list = Crystal::PointerLinkedList(Receiver(T)).new
+
     @lock.sync do
       @closed = true
 
-      @senders.each do |sender|
-        sender.state_ptr.value = DeliveryState::Closed
-        select_context = sender.select_context
-        if select_context.nil? || select_context.try_trigger
-          sender.fiber.enqueue
-        end
-      end
+      @senders, sender_list = sender_list, @senders
+      @receivers, receiver_list = receiver_list, @receivers
+    end
 
-      @receivers.each do |receiver|
-        receiver.state_ptr.value = DeliveryState::Closed
-        select_context = receiver.select_context
-        if select_context.nil? || select_context.try_trigger
-          receiver.fiber.enqueue
-        end
+    sender_list.each do |sender_ptr|
+      sender_ptr.value.state = DeliveryState::Closed
+      select_context = sender_ptr.value.select_context
+      if select_context.nil? || select_context.try_trigger
+        sender_ptr.value.fiber.enqueue
       end
+    end
 
-      @senders.clear
-      @receivers.clear
+    receiver_list.each do |receiver_ptr|
+      receiver_ptr.value.state = DeliveryState::Closed
+      select_context = receiver_ptr.value.select_context
+      if select_context.nil? || select_context.try_trigger
+        receiver_ptr.value.fiber.enqueue
+      end
     end
```

Comment:
You could iterate both arrays in one go by combining them `(sender_list + receiver_list).each do |ptr|`

Reply:
That allocates a new array and if this is a hot path it will slow down things.

Reply:
You could rewrite it using a `Tuple` instead:

```cr
{sender_list, receiver_list}.each do |list|
  list.each do |ptr|
    # ...
  end
end
```

Reply:
Yes, or a macro to unroll the loop

Reply:
Or just a private helper method for the iteration body?

Reply:
@straight-shoota Like this?
```crystal
sender_list.each(&.value.close_helper)
receiver_list.each(&.value.close_helper)
```